packages/lisk-transaction-pool/src/transaction_pool.ts

```diff
@@ -40,62 +40,136 @@ interface Queues {
+			}
+		});
+	}
 
-	public getProcessableTransactions(
-		limit: number,
-	): ReadonlyArray<Transaction> {}
+	public existsInTransactionPool(transaction: Transaction): boolean {
+		return Object.keys(this._queues).reduce(
+			(previousValue, currentValue) =>
+				previousValue || this._queues[currentValue].exists(transaction),
+			false,
+		);
+	}
 
-	public onDeleteBlock(block: Block): void {}
+	public get queues(): Queues {
+		return this._queues;
+	}
 
-	public onNewBlock(block: Block): void {}
+	public getProcessableTransactions(limit: number): ReadonlyArray<Transaction> {
+		return this._queues.ready.dequeueUntil(
+			queueCheckers.returnTrueUntilLimit(limit),
+		);
+	}
 
-	public onRoundRollback(delegates: ReadonlyArray<string>): void {}
+	public onDeleteBlock(block: Block): void {
+		const { received, validated, ...otherQueues } = this._queues;
 
-	public verifyTransaction(): void {}
+		// Move transactions from the verified, pending and ready queues to the validated queue where account was a receipient in the delete block
+		const transactionsToAffectedAccounts = this.removeTransactionsFromQueues(
+			otherQueues,
+			queueCheckers.checkTransactionForRecipientId(block.transactions),
+		);
 
-	private existsInTransactionPool(transaction: Transaction): boolean {}
+		this._queues.validated.enqueueMany(transactionsToAffectedAccounts);
+		// Add transactions to the verfied queue which were included in the deleted block
+		this._queues.verified.enqueueMany(block.transactions);
+	}
 
-	private expireTransactions(): void {}
+	public onNewBlock(block: Block): void {
+		// Remove transactions in the transaction pool which were included in the new block
+		this.removeTransactionsFromQueues(
+			this._queues,
+			queueCheckers.checkTransactionForId(block.transactions),
+		);
+
+		const { received, validated, ...otherQueues } = this._queues;
+		// Remove transactions from the verified, pending and ready queues which were sent from the accounts in the new block
+		const transactionsFromAffectedAccounts = this.removeTransactionsFromQueues(
+			otherQueues,
+			queueCheckers.checkTransactionForSenderPublicKey(block.transactions),
+		);
+
+		// Remove all transactions from the verified, pending and ready queues if they are of a type which includes unique data and that type is included in the block
+		// TODO: remove the condition for checking `containsUniqueData` exists, because it should always exist
+		const blockTransactionsWithUniqueData = block.transactions.filter(
+			(transaction: Transaction) =>
+				transaction.containsUniqueData && transaction.containsUniqueData(),
+		);
+		const transactionsOfTypesWithUniqueData = this.removeTransactionsFromQueues(
+			otherQueues,
+			queueCheckers.checkTransactionForTypes(blockTransactionsWithUniqueData),
+		);
+
+		// Add transactions which need to be reverified to the validated queue
+		this._queues.validated.enqueueMany([
+			...transactionsFromAffectedAccounts,
```

Comment:
Nice question! But I think these 2 arrays will always contain unique entries, this is because of the fact that `removeTransactionsFromQueues` *removes* transactions from the queues and not copies/clones them. Which means that if there is a transaction from an affected senderPublicKey and contains unique data, it will be removed from the queues by the `checkTransactionForSenderPublicKey` filter (which is executed first), and by the time `checkTransactionForTypes` filter is executed, it will not be in the queues.